<h1 align=center><font size = 5>Peer-graded Assignment: Build a Regression Model in Keras</font></h1>


Author: Konstantinos Kontogeorgopoulos

Final Assignment for Introduction to Deep Learning & Neural Networks with Keras, 
IBM Skills Network

In [34]:
#Install libraries if you do not already have them
#!pip install numpy==1.21.4
#!pip install pandas==1.3.4
#!pip install keras==2.1.6

## Lets first Download and clean the dataset

In [35]:
# We import the pandas and numpy libraries
import pandas as pd
import numpy as np

We read the dataset provided by the assignment https://cocl.us/concrete_data  
I downloaded it into the same folder as this FinalAssignment.ipynb

In [36]:
#read .csv data file and put it into concrete_data pandas dataframe object
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
#lets see the first 5 rows of the csv
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

This dataset is a concrete samples dataset.  
I see that i have Cement, Blast Furnace Slag, Fly ash, Water, Superplasticizer, Coarse Aggrepate, Fine Aggregate, Age and Strength as the columns

I check how many data points i have

In [37]:
concrete_data.shape

(1030, 9)

And i see that i have 1030 samples with the 9 columns that i talked about above.

Let's check if the data needs cleaning.

In [38]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [39]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

There are no NULL values in our data.  
So our data is clean and we do not need to clean them.

## Now that i know my data is clean i proceed to split the data into predictors and target according to the assignment's instructions

The target is Strength and all other columns are predictors

In [40]:
#Pass into the variable concrete_data_columns the concrete_data columns
concrete_data_columns = concrete_data.columns
#Pass into predictors variable the predictors, everything  except strength
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
#Pass into the target variable strength 
target = concrete_data['Strength']

And i doublecheck the predictors

In [41]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

And the target

In [42]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

## B: Normalize the Data 

Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.  
As the instructions prompt me to do.

In [43]:
# I normalize the data by subtracting the mean from the individual predictors and dividing by the standard deviation
predictors_norm = (predictors - predictors.mean()) / predictors.std()

# I print the top 5
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

## Build a Baseline Model

So lets import Keras using tensorflow 2.11

In [44]:
import tensorflow as tf 
# print(tf.version.VERSION)
# from tensorflow import keras

## D: Increase the number of hidden layers to 3

Use the Keras library to build a neural network with the following:

- Three hidden layers of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

In [45]:
# Regression model
def regression_model():
    #I create the model
    model = tf.keras.Sequential()
    # In Dense method i put in the hidden layer 10 neurons with the RelU activation function
    model.add(tf.keras.layers.Dense(10, activation='relu', input_shape=(8,)))
    # Add the 2 additional hidden layers to the neural network
    model.add(tf.keras.layers.Dense(10, activation='relu'))
    model.add(tf.keras.layers.Dense(10, activation='relu'))
    model.add(tf.keras.layers.Dense(1))
    
    # and add the adam optimizer and the mean squared error as the loss function 
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

In [46]:
#import train_test_split from sklearn
from sklearn.model_selection import train_test_split
#and split the data into a training and test sets with 70-30 % training - test
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=7)

2. Train the model on the training data using 50 epochs.

In [47]:
# I call the method to create the model and put it into model var
model = regression_model()
# and i use 50 epochs to train it
model.fit(X_train, y_train, epochs=50, verbose=1)

Epoch 1/50
23/23 [==============================] - 1s 864us/step - loss: 63406.5078
Epoch 2/50
23/23 [==============================] - 0s 773us/step - loss: 4754.4214
Epoch 3/50
23/23 [==============================] - 0s 818us/step - loss: 768.1228
Epoch 4/50
23/23 [==============================] - 0s 773us/step - loss: 645.5560
Epoch 5/50
23/23 [==============================] - 0s 818us/step - loss: 595.6786
Epoch 6/50
23/23 [==============================] - 0s 773us/step - loss: 581.8517
Epoch 7/50
23/23 [==============================] - 0s 773us/step - loss: 567.4668
Epoch 8/50
23/23 [==============================] - 0s 818us/step - loss: 553.7631
Epoch 9/50
23/23 [==============================] - 0s 773us/step - loss: 538.4346
Epoch 10/50
23/23 [==============================] - 0s 773us/step - loss: 525.7601
Epoch 11/50
23/23 [==============================] - 0s 773us/step - loss: 510.3493
Epoch 12/50
23/23 [==============================] - 0s 773us/step - loss: 498.990

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [48]:
#evaluate the model 
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 781us/step


204.93870544433594

In [49]:
# import mean_squared_error from sklearn.metrics
from sklearn.metrics import mean_squared_error
# compute the mean squared error between the predicted concrete stregth y_pred and the actual concrete strength y_test 
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

204.93871490573198 0.0


4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.  
5. Report the mean and the standard deviation of the mean squared errors.

In [50]:
all_mean_squared_errors = 50
#creating empty array
mean_squared_errors = []
#for 50 times i do the same as i did before
for i in range(0, all_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=50, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " + str(all_mean_squared_errors) + " mean squared errors with normalized data. \n")
print("Mean: "+ str(mean))
print("Standard Deviation: "+ str(standard_deviation))

MSE 1: 80.96833801269531
10/10 [==============================] - 0s 666us/step
MSE 2: 85.04658508300781
10/10 [==============================] - 0s 667us/step
MSE 3: 62.03527069091797
10/10 [==============================] - 0s 778us/step
MSE 4: 58.13543701171875
10/10 [==============================] - 0s 667us/step
MSE 5: 53.06267547607422
10/10 [==============================] - 0s 667us/step
MSE 6: 53.57664489746094
10/10 [==============================] - 0s 778us/step
MSE 7: 64.06069946289062
10/10 [==============================] - 0s 667us/step
MSE 8: 52.73115539550781
10/10 [==============================] - 0s 666us/step
MSE 9: 49.46168518066406
10/10 [==============================] - 0s 667us/step
MSE 10: 47.6562385559082
10/10 [==============================] - 0s 778us/step
MSE 11: 46.20335388183594
10/10 [==============================] - 0s 778us/step
MSE 12: 44.11347198486328
10/10 [==============================] - 0s 667us/step
MSE 13: 52.429134368896484
10/10 [====

## Question: How does the mean of the mean squared errors compare to that from Step B?

So at B the Mean was : 59.59618163562697 
and at D it is : 44.84507716417054
The Mean at D after we normalized our data is quite smaller than in question B where we did not.  
That means that after we added the extra two hidden layers the model has even less error than before.  
Actually we can see that this model(D) is the best one so far compared to (A),(B) and (C).  
